native/dlasq2.go

```diff
@@ -0,0 +1,330 @@
+	}
+
+	for k := 2 * n; k >= 2; k -= 2 {
+		z[2*k-1] = 0
+		z[2*k-2] = z[k-1]
+		z[2*k-3] = 0
+		z[2*k-4] = z[k-2]
+	}
+	i0 := 0
+	n0 := n - 1
+
+	if cbias*z[4*i0] < z[4*n0] {
+		ipn4Out := 4 * (i0 + n0 + 2)
+		for i4loop := 4 * (i0 + 1); i4loop <= 2*(i0+n0+1); i4loop += 4 {
+			i4 := i4loop - 1
+			ipn4 := ipn4Out - 1
+			z[i4-3], z[ipn4-i4-4] = z[ipn4-i4-4], z[i4-3]
+			z[i4-1], z[ipn4-i4-6] = z[ipn4-i4-6], z[i4-1]
+		}
+	}
+
+	pp := 0
+	for k := 0; k < 2; k++ {
+		d = z[4*n0+pp]
+		for i4loop := 4*n0 + pp; i4loop >= 4*(i0+1)+pp; i4loop -= 4 {
+			i4 := i4loop - 1
+			if z[i4-1] <= tol2*d {
+				z[i4-1] = math.Copysign(0, -1)
+				d = z[i4-3]
+			} else {
+				d = z[i4-3] * (d / (d + z[i4-1]))
+			}
+		}
+		emin = z[4*(i0+1)+pp]
+		d = z[4*i0+pp]
+		for i4loop := 4*(i0+1) + pp; i4loop <= 4*n0+pp; i4loop += 4 {
+			i4 := i4loop - 1
+			z[i4-2*pp-2] = d + z[i4-1]
+			if z[i4-1] <= tol2*d {
+				z[i4-1] = math.Copysign(0, -1)
+				z[i4-2*pp-2] = d
+				z[i4-2*pp] = 0
+				d = z[i4+1]
+			} else if safmin*z[i4+1] < z[i4-2*pp-2] && safmin*z[i4-2*pp-2] < z[i4+1] {
+				tmp := z[i4+1] / z[i4-2*pp-2]
+				z[i4-2*pp] = z[i4-1] * tmp
+				d *= tmp
+			} else {
+				z[i4-2*pp] = z[i4+1] * (z[i4-1] / z[i4-2*pp-2])
+				d = z[i4+1] * (d / z[i4-2*pp-2])
+			}
+			emin = math.Min(emin, z[i4-2*pp])
+		}
+		z[4*(n0+1)-pp-3] = d
+		qmax = z[4*(i0+1)-pp-3]
+		for i4loop := 4*(i0+1) - pp + 2; i4loop <= 4*(n0+1)+pp-2; i4loop += 4 {
+			i4 := i4loop - 1
+			qmax = math.Max(qmax, z[i4])
+		}
+		pp = 1 - pp
+	}
+
+	var ttype int
+	var dmin1, dmin2, dn, dn1, dn2, g, tau float64
+	var tempq float64
+	iter := 2
+	var nFail int
+	nDiv := 2 * (n0 - i0)
+	var i4 int
+	for iwhila := 1; iwhila <= n+1; iwhila++ {
+		if n0 < 0 {
+			goto OneSeventy
+		}
+		var desig float64
+		var sigma float64
+		if n0 != n-1 {
+			sigma = -z[4*(n0+1)-2]
+		}
+		if sigma < 0 {
+			info = 1
+			return info
+		}
+		var emax float64
+		if n0 > i0 {
+			emin = math.Abs(z[4*(n0+1)-6])
+		} else {
+			emin = 0
+		}
+		qmin := z[4*(n0+1)-4]
+		qmax = qmin
+		var out bool
+		for i4loop := 4 * (n0 + 1); i4loop >= 8; i4loop -= 4 {
+			i4 = i4loop - 1
+			if z[i4-5] <= 0 {
+				out = true
+				break
+			}
+			if qmin >= 4*emax {
+				qmin = math.Min(qmin, z[i4-3])
+				emax = math.Max(emax, z[i4-5])
+			}
+			qmax = math.Max(qmax, z[i4-7]+z[i4-5])
+			emin = math.Min(emin, z[i4-5])
+		}
+		if !out {
+			i4 = 3
+		}
+		i0 = (i4+1)/4 - 1
+		pp = 0
+		if n0-i0 > 1 {
+			dee := z[4*i0]
+			deemin := dee
+			kmin := i0
+			for i4loop := 4*(i0+1) + 1; i4loop <= 4*(n0+1)-3; i4loop += 4 {
+				i4 := i4loop - 1
+				dee = z[i4] * (dee / (dee + z[i4-2]))
+				if dee <= deemin {
+					deemin = dee
+					kmin = (i4+4)/4 - 1
+				}
+			}
+			if (kmin-i0)*2 < n0-kmin && deemin <= 0.5*z[4*n0] {
+				ipn4Out := 4 * (i0 + n0 + 2)
+				pp = 2
+				for i4loop := 4 * (i0 + 1); i4loop <= 2*(i0+n0+1); i4loop += 4 {
+					i4 := i4loop - 1
+					ipn4 := ipn4Out - 1
+					z[i4-3], z[ipn4-i4-4] = z[ipn4-i4-4], z[i4-3]
```

Comment:
This block illustrates why 0-based indexing is superior. The offsets in FORTRAN mean different things depending on the context.